boards/arm/mimxrt1060_evk/mimxrt1060_evk-pinctrl.dtsi

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2022, NXP
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Generated by rt_cfg_utils.py on 2022-01-27
+ */
+
+#include <nxp/nxp_imx/rt/mimxrt1062-iomuxc.h>
+
+&pinctrl {
+	lpuart1_default: lpuart1_default {
+		iomuxc_gpio_ad_b0_12_lpuart1_tx: iomuxc_gpio_ad_b0_12_lpuart1_tx {
+			pinmux = <IOMUXC_GPIO_AD_B0_12_LPUART1_TX>;
+			bias-bus-hold;
+			input-enable;
+		};
+
+		iomuxc_gpio_ad_b0_13_lpuart1_rx: iomuxc_gpio_ad_b0_13_lpuart1_rx {
+			pinmux = <IOMUXC_GPIO_AD_B0_13_LPUART1_RX>;
+			bias-bus-hold;
+		};
+
+	};
+
+	sai1_default: sai1_default {
+		iomuxc_gpio_ad_b1_09_sai1_mclk: iomuxc_gpio_ad_b1_09_sai1_mclk {
+			pinmux = <IOMUXC_GPIO_AD_B1_09_SAI1_MCLK>;
+			bias-bus-hold;
+			input-enable;
+		};
+
+		iomuxc_gpio_ad_b1_12_sai1_rx_data00: iomuxc_gpio_ad_b1_12_sai1_rx_data00 {
+			pinmux = <IOMUXC_GPIO_AD_B1_12_SAI1_RX_DATA00>;
+			bias-bus-hold;
+			input-enable;
+		};
+
+		iomuxc_gpio_ad_b1_13_sai1_tx_data00: iomuxc_gpio_ad_b1_13_sai1_tx_data00 {
+			pinmux = <IOMUXC_GPIO_AD_B1_13_SAI1_TX_DATA00>;
+			bias-bus-hold;
+			input-enable;
+		};
+
+		iomuxc_gpio_ad_b1_14_sai1_tx_bclk: iomuxc_gpio_ad_b1_14_sai1_tx_bclk {
+			pinmux = <IOMUXC_GPIO_AD_B1_14_SAI1_TX_BCLK>;
+			bias-bus-hold;
+			input-enable;
+		};
+
+		iomuxc_gpio_ad_b1_15_sai1_tx_sync: iomuxc_gpio_ad_b1_15_sai1_tx_sync {
+			pinmux = <IOMUXC_GPIO_AD_B1_15_SAI1_TX_SYNC>;
+			bias-bus-hold;
+			input-enable;
+		};
```

Comment:
you're not applying grouping principle here, you should have:

```
sai1_default: sai1_default {
	pinmux = <IOMUXC_GPIO_AD_B1_09_SAI1_MCLK>,
		        <IOMUXC_GPIO_AD_B1_12_SAI1_RX_DATA00>,
		        <IOMUXC_GPIO_AD_B1_13_SAI1_TX_DATA00>,
		        <IOMUXC_GPIO_AD_B1_14_SAI1_TX_BCLK>,
		        <IOMUXC_GPIO_AD_B1_15_SAI1_TX_SYNC>;
	bias-bus-hold;
	input-enable;
};
```

Reply:
I don't think it it possible to apply grouping for iMX.RT parts using the pinctrl API. Each pinmux option is defined in a header file as something like this:
`#define IOMUXC_GPIO_AD_B1_15_SAI1_TX_SYNC 0x401f8138 3 0x401f85ac 1 0x401f8328`

Those values correspond to the following registers in the datasheet:
- Pin specific `SW_MUX_CTL_PAD` register (0x401f8138 in this example)
- Pin specific value to write to `MUX_MODE` field of the `SW_MUX_CTL_PAD` register (3)
- Peripheral specific `SELECT_INPUT` register (0x401f85ac)
- Value to write to `DAISY` field of `SELECT_INPUT` register (1)
- Pin specific `SW_PAD_CTL_PAD` register (0x401f8328)

All of these values are unique to either the pin or peripheral being selected, so they can't be shared. The issue I run into with pin grouping is that DTS does not have a way to define an "array of arrays", which is what the pinmux option in pin grouping is. I haven't been able to find a workaround for this that does not require modifying the pinctrl API.

Since the maximum offset from the base address of `0x401f8000` requires 10 bits to store, there is no way to pack three register offsets, plus two 3 bit register settings, into a 32 bit integer, the way that implementations like #39974 do.

The only solutions I see so far are these two:
- Do not support pin grouping
- Encode each pinmux option as an array of `phandles`, and store the pinmux data itself in the dts node the `phandle` references.

Do you have any feedback on which option would work better, or is there a trick I'm missing? I'd prefer to use pin grouping, but I'm unsure how to enable it given this issue

Reply:
Have you considered encoding the offsets instead, maybe:

```c
/* Encode:
 * bits 0..11: Offset, used to resolve SW_MUX_CTL_PAD_* and SW_PAD_CTL_PAD_*
 *                  Looks like SW_PAD_CTL_PAD_* = SW_MUX_CTL_PAD_ + 0x200U
 * bits 12..23: 
 * bits 24..26: MUX_MODE (0..7)
 * bit 27: DAISY
 */
#define IOMUXC_GPIO_AD_B1_15_SAI1_TX_SYNC (0x138U | (0x5acU << 12U) | (3U << 24U) | (1U << 27U)
```

I'm not familiar enough with your platform to judge if that would scale to all your SoCs.

Reply:
You could also look at encoding _distances_, for example:

Any `IOMUX_SW_MUX_*` ranges from 0x14 to 0x200, 9 bits
Any `IOMUX_SW_PAD_*` ranges from 0x204 to 0x3f0, 9 bits
Any `IOMUX_*_SELECT_INPUT` ranges from 0x3f4 to 0x78c, 10 bits

That leaves 4 bits, 3 for MUX and 1 for DAISY. Again, it's really pushing the limits and may not be exportable to all your SoCs...

Reply:
Another trick I'm thinking of: Encode offsets in words, not bytes.

`IOMUX_SW_MUX_*` can be stored as: (offset - 0x14) / 4. Range of values to save: 0...0x7b (7 bits)
`IOMUX_SW_PAD_*` can be stored as: (offset - 0x204) / 4. Range of values to save: 0...0x7b (7 bits)
`IOMUX_*_SELECT_INPUT` can be stored as: (offset - 0x3f4) / 4. Range of values to save: 0...0x398 (10 bits)

So you have now: 7 + 7 + 10 + 3 + 1 = 28 bits, you are safe to even save some bits.

Then we'd have something like:
```c
#define IOMUXC_GPIO_AD_B1_15_SAI1_TX_SYNC (0x49U | (0x49U << 7U) | (0xEA << 14U) | (3U << 24U) | (1U << 27U)
```

Reply:
This could work. The `0x14` and `0x204` offsets can be stored in the DTS for the `pinctrl` itself, so those being SOC specific would be a nonissue. My only concern is that some pins are configured via a different peripheral (`IOMUXC_SVNS`), with a different base address. I'll have to figure out a way to indicate that difference to the driver, but it shouldn't take more than a few bits.

I'll look at reference manuals for the remainder of the iMX.RT line and see if this would work for higher pin count SOCs, and if so implement it. Thanks for the idea!